### DISCUSSION TITLE
Got git exception when using Git-Sync sidecar with Persistence enabled

### DISCUSSION THREAD

**Discussion by mis98zb** [apache/airflow#32146] [General] [answered]
### Official Helm Chart version

1.9.0 (latest released)

### Apache Airflow version

2.6.2

### Kubernetes Version

1.24.9

### Helm Chart configuration

```yaml
dags:
  persistence:
    # Annotations for dags PVC
    annotations: {}
    # Enable persistent volume for storing dags
    enabled: true
    # Volume size for dags
    size: 1Gi
    # If using a custom storageClass, pass name here
    storageClassName: azurefile-csi-premium
    # access mode of the persistent volume
    accessMode: ReadWriteMany
    ## the name of an existing PVC to use
    existingClaim:
    ## optional subpath for dag volume mount
    subPath: ~
  gitSync:
    enabled: true
```

### Docker Image customizations

none

### What happened

After deploy, scheduler pod got coninuous failure:
```
kubectl logs airflow-int-scheduler-5c849b586c-m2t2f -n airflow-int -c git-sync-init
INFO: detected pid 1, running init handler
I0626 10:04:23.831285      12 main.go:401] "level"=0 "msg"="starting up" "pid"=12 "args"=["/git-sync"]
I0626 10:04:23.847016      12 main.go:950] "level"=0 "msg"="cloning repo" "origin"="https://▩▩▩▩▩▩▩▩▩▩▩▩▩▩▩/PipelineAirflowDag" "path"="/git"
E0626 10:04:24.851556      12 main.go:547] "msg"="too many failures, aborting" "error"="Run(git clone -v --no-checkout -b Shadow --depth 1 https://▩▩▩▩▩▩▩▩▩▩▩▩▩▩▩/PipelineAirflowDag /git): exit status 128: { stdout: "", stderr: "Cloning into '/git'...\nerror: chmod on /git/.git/config.lock failed: Operation not permitted\nfatal: could not set 'core.filemode' to 'false'" }" "failCount"=1
```
git-sync-int cannot clone repo to dags volume.

### What you think should happen instead

git-sync-int should be able to clone the repo without exception.

### How to reproduce

deploy the helm file to azure kubernetes service.

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by potiuk** [comment]
I believe git-sync + persistence is a generally bad combination. Why would you want to use both git-sync and persistence? It has a number of problems if you try to do so. For example git-sync will generally create a full copy of checked out code every time it runs a new check-out and then it swaps the symbolic links, which causes a number of problems with persistence - because most of the persistence solutions do not know how to handle symbolic link swap and what is going to happen is they are syncing the whole change  in a non-atomic way right after git-sync syncs the files - which can cause all kind of problems when the files are both copied by persistence and swapped by git-sync at the same time.

 I wrote about it in this article in more detail https://medium.com/apache-airflow/shared-volumes-in-airflow-the-good-the-bad-and-the-ugly-22e9f681afca

I converted it into dicussion because I would love to hear from those who attempt to use both git-sync and persistence why git-sync is not enough for them. Why would you think you need both ? Why git-sync without persisten is not enough for you?

**Reply by potiuk** [reply]
BTW. Likely the filesytem your persistence use simply won't support git-sync way of accessing the filesystem (it's likely not a POSIX-compliant filesystem, likely some windows-specific one that you use), so anyhow it's not fixable until you change your filesystem to use persistence mechanism that git-sync will be somewhat compatible with (but I still would love to see why persistence in this case would be needed).

**Reply by mis98zb** [reply]
Previously I used git-sync without persistence, so all trigger/scheduler/worker pods got their own git-sync containers.
It worked well.
But with the increasing of replica, a lot of requests were sent to our git repo, so as my server ip is blocked by git server.... 
And when git-sync restarted due to the access failure, my airflow components (for example, web ui) also lose connection...
T_T

One solution is to enlarge the interval of git-sync to reduce the traffic, but in helm value file there is a comment saying that large interval will have a risk of different DAG definitions among airflow components.... 
T_T*2

So I can only try to use git-sync + persistence to reduce the traffic and get rid of the gap among components....

And my topology seems a little different from those in your article, :)
> - In below graph, the arrow direction is the dag transport direction.
> - In the pure_dag_repo, there is nothing except dag files.
> - Only git-sync write the shared volume, other airflow components all read the shard volume. No external access for the shared volume.
> - Currently the git-sync sidecar is running in scheduler pod, but I think it should be better that git-sync sidecar running in dag processor pod. However, according to helm chat 1.9.0, only without persistence the git-sync sidecar can be run in dag processor pod. What a pity. 
```mermaid
graph TB
dag_author_a --git--> pure_dag_repo
dag_author_b --git--> pure_dag_repo
pure_dag_repo --git-sync--> PVC
PVC --> airflow_dag_processor
PVC --> airflow_scheduler
PVC --> airflow_trigger
PVC --> airflow_worker_a
PVC --> airflow_worker_b
```
Can you help to comment for this pattern?

**Reply by potiuk** [reply]
It's the same. git-sync symbolic link swap will also cause enormous amount of traffic and lack of atomicity for the updates on your workers/triggerers/processors so my article remains very valid.

What is your frequency of polling ? I can't think of reasonably exceeding any kind of limits for the git repo with such an architecture even for big installation. What were your rate limits and numbers? This is quite surprising to see authenticated git requests even from multiple servers ? Maybe you were using public repos and unauthenticated requests ? because I think this is the only case (at least in GitHub) where IP server is blocked (and there the limits are low). Usually rate-limitng (much higher) is per authenticated user - so maybe there you had some problems?

**Reply by mis98zb** [reply]
The auth is ok.
We are using azure devops, a kind of copy of GitHub, :D

The sync frequency is default, 5 second. And we have 12 git-sync instances querying the changes.
I guest may be there is some anti-hack rule triggered by our pooling...
One more possibility may be the network is not stable.
Whatever, I don't want git-sync failure impact airflow component, so if it can be running only in dag proccessor pod, it will be perfect...

I know there is a way to set a failure limit before git-sync abort.
But without reducing the traffic, this tolerance may increase the risk of inconsistent DAG file among airflow components...

BTW, since the dag is already serialized to DB, why schduler/trigger still need read DAG files?
And, is it the best way to share DAG file via DB (maybe no sql db) with a standalone dag processor?
It can provide best consistency, and have no worry about environment support like file system.
Even get connection issue with git repo, it will not impact airflow components like scheduler/trigger/worker...

**Reply by potiuk** [reply]
Git syc failure should not absolutely lead to failure of the component. I think Git-sync has built in features to be resilient to temporary errors and as a side-car it should not impact runnnig services.  

Also Git-sync has the capability on running sync on signal not only timeout (so for example you could have long timout + you could  send signals to all your git-sync sidecars (and have a webhook in your azure git to notify and produce the signals) - that woulld save a lot of polling and would get nearly instant synchronisation. This is an easy optimization that you can make if you are worried about delays.

> But without reducing the traffic, this tolerance may increase the risk of inconsistent DAG file among airflow components...

You already have it. while you might think persistence will give you instant synchronization of all DAGs to all the components - this is wrong perception (and would be short of magical if it works) - when you have persistence and the files are updated in one place (where git-sync is running), the updated DAG files in the other "locations" will also be visible after some time. and not only that - they might be inconsistent. As opposed to git-sync, you might see different versions of different files at the same time, because only one of the updated files has already been synced across the filesystem. And even worse you can see the different status of different files in different workers. 

So you have basically the same as you have in git-sync -only worse, becuase git-sync at least will manage atomicity between different files. Imagine you added a function in one common library file and use it in a DAG. the lack of atomicity might mean that the DAG will run the updated method will call the common library method and it will be missing there because the common library method that the worker sees is the previous version of it.  So you get even worse kinds of problems when you use persistency.

> BTW, since the dag is already serialized to DB, why schduler/trigger still need read DAG files?
> And, is it the best way to share DAG file via DB (maybe no sql db) with a standalone dag processor?
> It can provide best consistency, and have no worry about environment support like file system.
> Even get connection issue with git repo, it will not impact airflow components like scheduler/trigger/worker...

Serialization is only structure of the DAG not Python code. Serialized DAG is Json that describes relations between tasks and objects. As mentioned above, your DAG might import another file, that file might import another etc. etc. Effectively you would have to store the whole DAG content in the database when you serialize DAG. Syncing the folder instead of storing the whole content of the folder is much more efficient way.

**Reply by mis98zb** [reply]
Thank you for the hint, I will try signal way on git-sync.

And, for file system in aks/kubernetes, my understanding is there is no local copy among containers, all data stored in provision server. Provision server may have some internal replicas, but it shows a consistent view to external users. This just like a database cluster.
```mermaid
sequenceDiagram

actor process_in_container

process_in_container ->> storage_volume: posix call for read/write
storage_volume ->> csi_driver: call
csi_driver ->> provision: adpated call for underlying file system
provision ->> provision: read/write ddata
```
So when git-sync update files, there should be no older copy in other "locations".
If the provision can support swap operation correctly, everything should be ok.
We are using files.csi.azure.com which use SMB3.0/3.1. I'll try to find if it supports swap correctly.

Please correct me if I have some misunderstanding...

**Reply by mis98zb** [reply]
> > BTW, since the dag is already serialized to DB, why schduler/trigger still need read DAG files?
> > And, is it the best way to share DAG file via DB (maybe no sql db) with a standalone dag processor?
> > It can provide best consistency, and have no worry about environment support like file system.
> > Even get connection issue with git repo, it will not impact airflow components like scheduler/trigger/worker...
> 
> Serialization is only structure of the DAG not Python code. Serialized DAG is Json that describes relations between tasks and objects. As mentioned above, your DAG might import another file, that file might import another etc. etc. Effectively you would have to store the whole DAG content in the database when you serialize DAG. Syncing the folder instead of storing the whole content of the folder is much more efficient way.

This makes sense for worker. But why scheduler/trigger need the whole code? The serialized DAG may have enough information for scheduling/triggering...

**Reply by potiuk** [reply]
> This makes sense for worker. But why scheduler/trigger need the whole code? The serialized DAG may have enough information for scheduling/triggering...

Scheduler doesn't need. Technically DAG files are needed for DagFileProcessor to parse and serialize the DAGs - this is by default part of scheduler, but you can separate it out to a standalone processor, and then scheduler does not need DAGs and you should not have DAGs available for Scheduler when you are using standalone file processor.

Triggerer also runs user code provided by the user - namely Triggers. If you place your Triggers in your DAGs folder, they will be loaded and executed from there. Not sure where your notion is that you do not need to have access to the Triggers by triggerer.

**Reply by potiuk** [reply]
> Please correct me if I have some misunderstanding...

There are few points that things can get desynchronized. And it is entierely dependning on the way how internal implementation of the distribution mechanism is done. 

There is a time between you rename/move/save files, flush them to disks, and the file gets distributed. Each distributed file system does it in a different way. And each of them has different characteristics there. So effectively at any point in time on the client side you see only those changes that a) has been saved b) has been flushed to disk in the node they were changed c) the file change information has been disstributed to the remote client (here - depending on the filesysystem this might or might not be somewhat synchronized - if you read my article carefully I was discussing history how such synchronisation was implemented in various versions of NFS in the past and there are various parameters you can use to control it - but there is always a trade-off of consistency vs. speed.  That's why we cannot make any assumption in "general" for distributed filesystem when it comes to consistency. But even flushing to disk on it's own is enough to understand that with sufficient number of changes, what you see locally as "saved" is not yet "flushed" because each filesystem is flushed independently. 

And yes if you had a filesystem that instead of copying files, understand symbolic link changes, that could be optimised and improved. But I am afraid in most cases, symbolic link swap is not something that distributed filesystem have support for when it comes to optimisaiton - simply becasue symbolic link is a higher-level construct than filesystem operations that are usually handled by such distributed filesystem. Symbolic link - from the filesystem perspective is just a file that points elsewhere and usually is implemented as a layer above the filesystem operations that distributed filesystem tap into when it comes to dirstributing changes. And even if you do - that's a lot of traffic for every sync with git-sync because effectively how git-sync works is that every time it checks out a new commit, it deletes the old version of the checked out sources, recreates them anew and then does the swap. So from distributed filesystem point of view - all the files checked out are new and has to be distributed to all the clients reading them (no caching involved) and when you swap them, usually what happens those files are sent again - because the symbolic link swap is not handled in optimised way and those files are **again** seen as new files.

And even if you do have a distributed filesystem that supports optimisations on the symbolic link swaps this is not all.

If you know how filesystem interface works - there is usually "file-by-file" work, not a "committ operation work". The problem is that when you have a "filesystem API" there is basically no mechanism to say "this is an atomic operation that impacts all the files" on the client side. The big exception to that is symbolic link swap on a filesystem and it only works for Python importing the files, because when you do symbolic link swap and Python is in the middle of importing of a module, it will consistently keep a handle of previously imported module even if the symbolic links gets swapped while it is importing files from any of the subdirectories of the module. In other words, when the top level folder has been already read by a parser, it will continue using that parsing from old location even if physically the file is in a different location:

So if you have `/base/DAGS -> /tmp/old_cached_copy/DAGS` and swap it to `/base/DAGS -> /tmp/new_copy/DAGS` and Python worker or DAG file processor started to parse Python files from `/tmp/old_cached_copy/DAGS` it will continue to import files from there because for Python interpreter, the handle for "/base/DAGS" still in the running Python fork interpreter points to "/tmp/old_cached_copy/DAGS" - this is how POSIX local filesystem works, and this is why swapping symbolic links is atomic operation for Python interpreter parsing not only single file but also other imported files from the same "DAG folder" version. In other words symbolic link swapping on local filesystem that `git-sync` is doing provides equivalent of snapshotting all the files that are imported by the DAG parsing during that parsing operation. 

This is nowhere near what remote filesystem swap can provide. It's physically impossible to take such a snapshot for remote system while it is doing the parsing - because generally speaking there is no built-in mechanism (and Python does not use it) to take a snapshot of such a filesystem and then do parsing. If you start replacing the files while parser is running, the client side does not have the snapshot of previous version of symbolically linked folder - all that it sees is that the files in the new location changed. And thus, your parsing might read different version of the files.

Of course there might be distrubuted filesystem that taps into this high-level mechanism of symbolic links swap and replicate it. But I think there is no FS that does it currently.

But If you find one and prove it does, I am happy to make exception in my article for that particular FS

**Reply by potiuk** [reply]
> If the provision can support swap operation correctly, everything should be ok.

BTW. Your filessystem does not support chmod. Do you realistically think the above behaviour and atomicity/snapshotting is something it supports ?

**Reply by mis98zb** [reply]
> Triggerer also runs user code provided by the user - namely Triggers. If you place your Triggers in your DAGs folder, they will be loaded and executed from there. Not sure where your notion is that you do not need to have access to the Triggers by triggerer.

Got it.
I have a misunderstanding for Triggerer functionality, thank you very much for correcting me!

I have an idea to enable dag processor + git-sync + persistence, and let git-sync sidecar running in dag processor pod.
This needs to change helm file slightly. Because when I enable all of dag processor + git-sync + persistence, git-sync sidecar does not appear in either scheduler pod or dag processor pod.

Git-sync sync DAG files to pvc, dag processor serialize DAG to db. Other components read DAG from db, read source code from pvc.
Do you think this workable logically?

**Reply by potiuk** [reply, marked as answer]
Again. The symbolic link swap will not provide you the atomicity, so I think this is a worse solution that having git-sync. Nothing changed vs. my article. I was actually even pondering disabling of the git-sync + persistence combo, but rather than that I decided to leave it for the user to decide if they want to open to the potential inconsistency or not. I would not recommend that solution because of that (on top of usually paying a lot because most of the PVCs like that have additional distribution lags - for example in AWS you have to pay extra IOPS (and a lot ) usually to handle the traffic that airflow generates.

But if you think that solution wil be ok and you accept all the tradeoffs and problem after the discussion - it's your choice eventually. That 's why I decided not to disable the combo, you as deployment manager should make the choices that are best for you based on the knowledge you have.

**Reply by mis98zb** [reply]
> So if you have /base/DAGS -> /tmp/old_cached_copy/DAGS and swap it to /base/DAGS -> /tmp/new_copy/DAGS and Python worker or DAG file processor started to parse Python files from /tmp/old_cached_copy/DAGS it will continue to import files from there because for Python interpreter, the handle for "/base/DAGS" still in the running Python fork interpreter points to "/tmp/old_cached_copy/DAGS" - this is how POSIX local filesystem works, and this is why swapping symbolic links is atomic operation for Python interpreter parsing not only single file but also other imported files from the same "DAG folder" version. In other words symbolic link swapping on local filesystem that git-sync is doing provides equivalent of snapshotting all the files that are imported by the DAG parsing during that parsing operation.

Yes, linux use file path to get i-node, and use i-node to do file operations. So after getting i-node, the file path change will not impact the file operation.
This is a chance for csi driver to bind i-node to file id.
As far as I know, SeaweedFS is manage files by uuid, and the file path is just a "decorater" supported by Filer Service.
It also has CSI driver, however, I did not use this driver.

**Reply by mis98zb** [reply]
> I would not recommend that solution because of that (on top of usually paying a lot because most of the PVCs like that have additional distribution lags - for example in AWS you have to pay extra IOPS (and a lot ) usually to handle the traffic that airflow generates.

Understood.
Yes, git-sync polls new commit, while dag processor scans all files.
File scan traffic is significantly much more than git commit polling.

Maybe it can be a optimization that use git-sync web hook to trigger dag processor reload the DAG files...

**Comment by mis98zb** [comment]
> BTW. Likely the filesytem your persistence use simply won't support git-sync way of accessing the filesystem (it's likely not a POSIX-compliant filesystem, likely some windows-specific one that you use), so anyhow it's not fixable until you change your filesystem to use persistence mechanism that git-sync will be somewhat compatible with (but I still would love to see why persistence in this case would be needed).

From the error log, I think it is git not git-sync which failed:
```
Run(git clone -v --no-checkout -b Shadow --depth 1 https://▩▩▩▩▩▩▩▩▩▩▩▩▩▩/PipelineAirflowDag /git): exit status 128: 
stderr: 
    Cloning into '/git'...
    error: chmod on /git/.git/config.lock failed: Operation not permitted
    fatal: could not set 'core.filemode' to 'false'
```

**Reply by potiuk** [reply]
Yes. git-sync runs git command that failed.

**Reply by mis98zb** [reply]
Yup, it seems the file system not support chmod, or the folder is mounted with wrong mode.
I checked the mounted folder on airflow components other than scheduler, the mode is 777, which should be ok...

**Comment by mis98zb** [comment]
Azure File shares got 62% fail rate in pjdfstest...
The failures can be classified as below:
![image](https://github.com/apache/airflow/assets/3025315/4cdd58f9-44b2-47cd-919a-c1742543c11e)
The shared folder and file has mode 777, owner is root, cannot be changed.
And support neither hard link nor symbol link.
T______T

I stop thinking about persistence....

**Reply by potiuk** [reply]
> I stop thinking about persistence....

Told ya.

I also captured some of the discussion in this PR: https://github.com/apache/airflow/pull/32181  - I think I answered the same questions and came to the same conclusions already several times, so maybe having it explicit in our docs will help others to get to the "I stop thinking about persistence..." decision way faster :).

**Reply by mis98zb** [reply]
btw, the new object store may have better support for posix APIs.
I found JuiceFS passed all pjdfstest.
SeaweedFS is also a object store.
Maybe that kind of object store can be the candidate file system for airflow...

However, it is not originally supported on Azure Kubernetes Service, so I stop thinking about persistence...

Currently my main task is to support concurrency control for mapping on task group.
I have a task group support only solution, and works okay on our environment.
But I want to get a unify solution to support both task and task group.
I create a PR #32122  and a discussion #31444 for it.
Please help to comment if possible, :)

**Reply by potiuk** [reply]
> Maybe that kind of object store can be the candidate file system for airflow...

FYI. We have no "selected" file-system (and will never have) and this is a deliberate choice - it's deployment manager's choice. We cannot dictate the user what kind of filesystem they use, this is part of the deployment of their choice. We can warn them to check POSIX compliance (and I think my PR is the most we can do), but seeting up, managing, fine-tuning is outside-of-the domain of Airlfow it's the role of Deployment Manager. We can at most give advices on what they can do. We cannot also be responsible for the choice - if say "this filesystem works well" it would have to be based on some more data and evidence. And we - as community with mostly volunteers who contribute in their free time - have no capacity or resources to make such checks and "vouch" for some solution (because it automatically falls on us if someone has problem with it and we have not checked). But there are managed services for Airlfow that make their choices and provide support - this is their role in the ecosystem, they provide "end-to-end" solution with paid support for deployment choices they made. For example - as you - some users have preferred deployment environments where things are supported this or that way (or not supported), so having a "selected" FS would also limit the ways how users could install it.

> I create a PR https://github.com/apache/airflow/pull/32122 and a discussion https://github.com/apache/airflow/discussions/31444 for it. Please help to comment if possible, :)

Separate issue - and tehre are others who know more there than I do.  If you do not get enough attention (usually we have 72H as a minimum time to expect any response from anyone), look at the history of the related features and ping those people who you think could help you. You can persistently ask there for support. Or better - since it looks like your issue is about adding significant conceptual change  (I am guessing) - you should raise it at dev@airflow.apache.org  - the discussion about some features that are more than small changes are usually happening there. See https://airflow.apache.org/community/ for details how to subscribe.
